typing/persistent_env.ml

```diff
@@ -401,23 +401,74 @@ let find_import ~allow_hidden penv ~check modname =
           add_import penv modname;
           acknowledge_import penv ~check modname psig
 
-let remember_global { globals; _ } global ~mentioned_by =
-  if Global_module.has_arguments global then
-    let global_name = Global_module.to_name global in
-    match Hashtbl.find globals global_name with
-    | exception Not_found ->
-        Hashtbl.add globals global_name
-          { gn_global = global; gn_mentioned_by = mentioned_by }
-    | { gn_global = old_global; gn_mentioned_by = first_mentioned_by } ->
-        if not (Global_module.equal old_global global) then
+let remember_global { globals; _ } global ~precision ~mentioned_by =
+  let mentioned_by =
+    match mentioned_by with
+    | `Other global_module -> global_module
+    | `Current ->
+        (* This leaves off the prefix, but we're only using this for reporting
+           what should be internal errors *)
+        CU.get_current_or_dummy ()
+        |> CU.to_global_name_without_prefix
+  in
+  let global_name = Global_module.to_name global in
+  match Hashtbl.find globals global_name with
+  | exception Not_found ->
+      Hashtbl.add globals global_name
+        { gn_global = (global, precision);
+          gn_mentioned_by = mentioned_by;
+        }
+  | { gn_global = old_global;
+      gn_mentioned_by = first_mentioned_by } ->
+      let new_global = global, precision in
+      match
+        Global_module.With_precision.meet old_global new_global
+      with
+      | Ok updated_global ->
+        if not (old_global == updated_global) then
+          Hashtbl.replace globals global_name
+            { gn_global = updated_global;
+              gn_mentioned_by = first_mentioned_by }
+      | Error Inconsistent ->
           error (Inconsistent_global_name_resolution {
               name = global_name;
               old_global;
-              new_global = global;
+              new_global;
               first_mentioned_by;
               now_mentioned_by = mentioned_by;
             })
```

Comment:
I'm unsure of this error - in what case would an user error cause this? I didn't find it in the test cases. If this must indicate an internal bug, it should raise instead?

Reply:
As discussed offline, this is now made into an internal error. Resolved.